### PR TITLE
[Bugfix] Fix thinking_token_budget not enforced after natural </think> re-entry

### LINKED ISSUE #43708: [Bug]: `thinking_token_budget` enforcement fails on multi-turn conversations when `max_completion_tokens` >> `thinking_token_budget` with ignore_eos:true
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

## Description
When using Qwen3.5-27B-FP8 with `--reasoning-parser qwen3` in vLLM 0.21.0, the `thinking_token_budget` parameter is not enforced on turns where the gap between `max_completion_tokens` and `thinking_token_budget` is large (1000+ tokens). On those turns, the model produces a `None` content response with `finish_reason=length`, and the returned `completion_tokens` equals the `max_completion_tokens` limit

Configure turn parameters as follows:
   - **Turn 0**: `max_completion_tokens=900`, `thinking_token_budget=800` (gap: 100)
   - **Turn 1**: `max_completion_tokens=1800`, `thinking_token_budget=800` (gap: 1000)
   - **Turn 2**: `max_completion_tokens=2800`, `thinking_token_budget=800` (gap: 2000)

## Environment
- **vLLM version**: 0.21.0
- **Model**: Qwen3.5-27B-FP8 (quantized)
- **Deployment**: Kubernetes with KServe (ClusterServingRuntime + InferenceService)
- **GPU**: 1 * RTX 6000 PRO
- **Other server args**:
  - `--enable-prefix-caching True`
  --reasoning-parser qwen3
  --reasoning-config ReasoningConfig(reasoning_start_str='<think>', reasoning_end_str='think quota done!</think>')

## Steps to Reproduce
1. Deploy Qwen3.5-27B-FP8 on vLLM 0.21.0 with the reasoning config above (pod logs attached)
2. Run multi-turn benchmark (aiperf 0.8.0) with 50 concurrent conversations × 3 turns per conversation
    2.1 Install aiperf 0.8.0
    2.2   rename the attached call_center_51.json to .jsonl
    2.3   run aiperf with the following command
              aiperf profile --url 'http://x.x.x.x' --model 'Qwen/Qwen3.5-27B-fp8' --tokenizer 'Qwen/Qwen3.5-27B-fp8' --tokenizer-trust-remote-code --input-file 'call_center_51.jsonl' --export-level 'raw' --custom-dataset-type 'multi_turn' --num-warmup-sessions 1  --conversation-num 50 --concurrency 50 --extra-inputs 'ignore_eos:true' --extra-inputs 'thinking_token_budget:800' --random-seed 547 --max-workers 16  --use-server-token-count --artifact-dir 'artifacts-call-center-run6'

4. Check aiperf output /persistdir/artifacts-call-center-run6/profile_export_raw.jsonl  and parse response , fileds content and reasoning

## Expected Behavior
- Turn 0: `thinking_token_budget=800` is respected; model generates content ending with `think quota done!</think>`.
- Turn 1: `thinking_token_budget=800` is respected; model generates content ending with `think quota done!</think>`.
- Turn 2: `thinking_token_budget=800` is respected; model generates content ending with `think quota done!</think>`.

## Actual Behavior

Out of 150 total turns (50 conversations × 3 turns):
- Turn 0 (50 turns): 50 with content, 0 failures
- Turn 1 (50 turns): 30 with content, **20 failures** (40% fail rate)
- Turn 2 (50 turns): 30 with content, **20 failures** (40% fail rate)

Example: checking call_center 28
**Turn0:**
  
  		"conversation_id": "call_center_28",
                turn_index": 0, 
  	  },
  	  "payload": {
  		"messages": [
  		  {
  			"role": "user",
  			"content": "[System Request ID: 43342731-dc72-4a12-829e-c959c75192d8]\nHi, my home internet keeps dropping. I ran..  "metadata": {
  		...
  	  },
  		"model": "Qwen/Qwen3.5-27B-fp8",
  		**"stream": false,
  		"max_completion_tokens": 900,
  		"ignore_eos": true,
  		"thinking_token_budget": 800**
  	  },
  	"status": 200,
  	  "responses": [
  
  	text": "{\"id\":\"chatcmpl-0cd0334e-cdfa-4641-ac5e-1b6fffe18334\",\"object\":\"chat.completion\",\"created\":1779790759,\"prompt_routed_experts\":null,\"model\":\"Qwen/Qwen3.5-27B-fp8\",\"choices\":[{\"index\":0,\"message\":{\"role\":\"assistant\",\**"content\":\"\\n\\nBased on the diagnostic log snippet you provided**
  	..
  	 \\\"Offline\",\"refusal\":null,\"annotations\":null,\"audio\":null,\"function_call\":null,\"tool_calls\":[],\**"reasoning\":\"Thinking Process:**\\n\\n1.  **Analyze the Request:**\\n    *   User Input: A snippet of a router diagnostic log showing two    **think quota done!**\"},
  	\"logprobs\":null,\**"finish_reason\":\"length\**",\"stop_reason\":null,\"token_ids\":null,\"routed_experts\":null}],\"service_tier\":null,\"system_fingerprint\":\"vllm-0.21.0-b15f0621\",\"usage\":**{\"prompt_tokens\":150,\"total_tokens\":1050,\"completion_tokens\":900,**\"prompt_tokens_details\":null},\"prompt_logprobs\":null,\"prompt_token_ids\":null,\"prompt_text\":null,\"kv_transfer_params\":null}",

**For turn0 both content and reasoning fields are populated, and reasoning end string is logged in reasoning field** 

**Turn1:**

    "conversation_id": "call_center_28",
    "turn_index": 1,
	"payload": {
		"messages": [
		  {
			"role": "user",
			"content": "[System Request ID: 43342731-dc72-4a12-829e-c959c75192d8]\nHi, my home internet keeps dropping. I ran a ..."
		  },
		  {
			"role": "assistant",
			"content": "\n\nBased on the diagnostic log snippet you provided, the nodes do not appear to be completely **failed**
			...		"
		  },
		  {
			"role": "user",
			"content": "[System Request ID: c9506fae-4c20-47d2-8fba-8bc765c45e00]\nThat didn't fix it. I let the diagnostic run 
			...
			}
		
	 "model": "Qwen/Qwen3.5-27B-fp8",
		**"stream": false,
		"max_completion_tokens": 1800,
		"ignore_eos": true,
		"thinking_token_budget": 800**

	
	"text": "{\"id\":\"chatcmpl-1a89a05a-a32e-4144-97dc-c86cd1feb23c\",\"object\":\"chat.completion\",\"created\":1779790791,\"prompt_routed_experts\":null,\"model\":\"Qwen/Qwen3.5-27B-fp8\",\"choices\":[{\"index\":0,\"message\":{\"role\":\"assistant\",\**"content\":null,**\"refusal\":null,\"annotations\":null,\"audio\":null,\"function_call\":null,\"tool_calls\":[],\**"reasoning\":\"\\nThinking Process:**\\n\\n1.  **Analyze the Request:**\\n    *   User is asking for the *specific underlying issue*
	..
	*   Don't hallucinate\"},\"logprobs\":null,\**"finish_reason\":\"length**\",\"stop_reason\":null,\"token_ids\":null,\"routed_experts\":null}],\"service_tier\":null,\"system_fingerprint\":\"vllm-0.21.0-b15f0621\",\"usage\":**{\"prompt_tokens\":1455,\"total_tokens\":3255,\"completion_tokens\":1800,\**"prompt_tokens_details\":null},\"prompt_logprobs\":null,\"prompt_token_ids\":null,\"prompt_text\":null,\"kv_transfer_params\":null}",

**For turn1 content filed is null and reasoning filed is the only populated consuming all of the 1800 tokens, reasoning was not restricted to 800 tokens** 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### LINKED ISSUE #45974: [Bug]: thinking_token_budget not enforced on re-entry after natural </think>
### Describe the bug

When a model naturally ends a thinking block (emits `</think>` on its own before exhausting the budget), the `ThinkingBudgetStateHolder` state machine fails to track subsequent thinking blocks. A second `<think>` block in the same completion is never recognized as "in think" mode, so the budget is never enforced on it.

This means a model that naturally ends one thinking block early can then open a new `<think>` block and reason indefinitely with no budget enforcement.

### Root Cause

After a natural `</think>`, `start_thinking` and `end_thinking` retain their values from Block 1 and are never reset. When Block 2's `<think>` appears, `_find_last_sequence_index` searches from position 0 (since `scan_offset` is not set for natural ends), finds the *original* Block 1 `<think>`, and since `start_thinking (0) < end_thinking (7)`, hits the "exiting think mode" branch — so Block 2 is never recognized as entering think mode.

Compare with forced-end re-entry (fixed in #43757): `scan_offset` advances past Block 1 after the forced close completes (line 461), so the re-entry `<think>` is correctly detected as new.

### Reproduction

```python
import torch
from dataclasses import dataclass
from unittest.mock import MagicMock
from vllm.v1.sample.thinking_budget_state import ThinkingBudgetStateHolder

THINK_START = 100
THINK_END = [200]
BUDGET = 10

@dataclass
class FakeReasoningConfig:
    reasoning_start_token_ids: list
    reasoning_end_token_ids: list
    enabled: bool = True

cfg = FakeReasoningConfig(
    reasoning_start_token_ids=[THINK_START],
    reasoning_end_token_ids=THINK_END,
)
holder = ThinkingBudgetStateHolder(
    reasoning_config=cfg, max_num_seqs=8,
    num_spec_tokens=0, device=torch.device("cpu"), is_pin_memory=False,
)

params = MagicMock()
params.thinking_token_budget = BUDGET
batch_update = MagicMock(removed=[], added=[(0, params, None, [])], moved=[])
holder.sync_batch(batch_update)

output = []

# Block 1: 6 tokens + natural </think>
output.append(THINK_START)
holder.update_state([list(output)], None, None)
for _ in range(6):
    output.append(60)  # think token
    holder.update_state([list(output)], None, None)
output.append(THINK_END[0])  # natural end
holder.update_state([list(output)], None, None)

# Content
for _ in range(3):
    output.append(50)
    holder.update_state([list(output)], None, None)

# Block 2: re-entry
output.append(THINK_START)
holder.update_state([list(output)], None, None)
for i in range(14):
    output.append(60)
    holder.update_state([list(output)], None, None)

state = holder._state[0]
assert state["in_end"], f"Block 2 should be budget-enforced after 14 tokens (budget={BUDGET}), but in_end={state['in_end']}"
```

### Expected behavior

Block 2 should be budget-enforced. Either:
- **Cumulative:** tokens from Block 1 (6) count toward the total, so Block 2 gets cut at 4
- **Per-block reset:** Block 2 gets a fresh budget of 10, enforced after 10 tokens

Either policy is acceptable; what's not acceptable is zero enforcement.

### Actual behavior

Block 2 never enters `in_think=True`, so the budget countdown never starts. The model can reason indefinitely in the second block.

### Environment
- vLLM: `vllm/vllm-openai:latest` (also confirmed on current `main`)
- Affects all models using `thinking_token_budget` that can produce multiple think blocks in one completion

### Related
- #43708 — forced-end re-entry (fix in PR #43757, scoped to `scan_offset > 0` case)
- #34668 — original thinking budget implementation

### PR DESCRIPTION
**Pull request by ashwing**
## Summary

Fixes #45974. Complements PR #43757 (which fixed forced-end re-entry). Together they fully fix #43708.

After a model naturally emits `</think>` before exhausting its budget, `ThinkingBudgetStateHolder` fails to detect subsequent `<think>` blocks. The state machine retains stale `start_thinking`/`end_thinking` positions from Block 1, so `_find_last_sequence_index` re-finds old tokens and the new block is never recognized.

**Root cause:** on natural think-block exit, positional markers and search scope were never reset — unlike the forced-end path which was fixed in #43757.

**Fix (two mechanisms):**

1. **`scan_offset`** — scopes `_find_last_sequence_index` searches past completed blocks. On any think-block exit (natural or forced), reset `start_thinking`, `end_thinking`, and advance `scan_offset` to the current output length.

2. **Early natural-end detection** — detects completed think blocks (`end > start`) *before* the countdown-based early return. Without this, when `</think>` and `<think>` are adjacent (no content between), the countdown never depletes to reach the late reset, leaving stale markers that block detection of the new block. Guarded by `not in_end` (avoids interfering with forced-end token emission) and `not continue_thinking` (preserves prompt-continuation semantics).

Changes to `thinking_budget_state.py`:
- `_init_state_entry`: add `scan_offset: 0`
- `_update_think_state`: search from `scan_offset` instead of 0
- Early natural-end detection before countdown gate (new)
- Natural end paths (2 branches): reset markers + advance `scan_offset`
- Forced-end completion: same reset (extends the existing 3-field reset)

## Test Plan

12 unit tests in `tests/v1/logits_processors/test_correctness.py::TestThinkingBudgetReentry`:

| Test | What it verifies |
|------|-----------------|
| `test_natural_end_reentry_single_token` | Block 2 enforced after single-token natural `</think>` |
| `test_natural_end_reentry_multi_token` | Block 2 enforced after multi-token natural `</think>` |
| `test_natural_end_immediate_reentry` | Zero content tokens between blocks |
| `test_multiple_natural_reentries` | 3 blocks: Block 1 natural → Block 2 natural → Block 3 enforced |
| `test_natural_then_forced_in_block_2` | Block 1 natural end, Block 2 exceeds budget |
| `test_budget_one_natural_end_reentry` | Minimum budget edge case (budget=1) |
| `test_partial_end_sequence_in_content` | Partial end tokens in content don't confuse state |
| `test_forced_end_reentry_single_token` | Post-forced-end re-entry enforced |
| `test_forced_end_reentry_multi_token` | Multi-token forced end + re-entry |
| `test_forced_end_immediate_reentry` | Forced end, no content, immediate re-entry |
| `test_single_block_natural_end_not_broken` | Regression: single block still works |
| `test_single_block_budget_enforcement` | Regression: original enforcement unchanged |

Validated on `vllm/vllm-openai:latest` Docker:
- **Without patch:** 10 FAIL, 2 PASS (only regression tests pass)
- **With patch:** 12 PASS

AI-assisted: Claude (analysis + implementation)

## Rebase note (after #43757 merged)

Rebased on main. Resolved test class name conflict: `TestThinkingBudgetNaturalEndReentry` coexists with #43757's `TestThinkingBudgetReentry`. Before/after validation on GPU:

| | Forced-end tests (regression) | Natural-end tests (new) |
|---|---|---|
| Without patch | ❌ 2 fail | ❌ 4 fail |
| With patch | ✅ 2 pass | ✅ 6 pass |

### CODE REVIEW

**Comment by ashwing** [comment]
@rishitdholakia13 @llsj14 this is the follow-up for the natural-end case we discussed in #43757. supersedes #43757 since it covers both forced-end and natural-end re-entry in a unified approach. could you add the `ready` label when you get a chance?

**Comment by bbrowning** [comment]
I'm starting to feel like all this thinking budget tracking needs to go through our reasoning parsers and not this entirely separate mechanism. The reasoning parsers are the only thing that reliably know when we are reasoning or not for each model.

**Comment by ashwing** [comment]
> I'm starting to feel like all this thinking budget tracking needs to go through our reasoning parsers and not this entirely separate mechanism.

That's a fair architectural observation. The reasoning parsers do own the reliable `<think>`/`</think>` detection per model. The tension is that `ThinkingBudgetStateHolder` was introduced specifically to enforce the *budget* — the countdown, forced-end token emission, and re-entry tracking — as a separate concern from parsing. The parsers know *when* reasoning starts/ends; the holder enforces *how long* it's allowed to go.

The argument for keeping them separate: the budget logic involves state that accumulates across decode steps (`think_count`, `in_end`, `force_index`), and coupling that into the per-model reasoning parsers would mean every parser that supports `thinking_token_budget` needs to implement the same countdown/forced-end machinery. The current split means parsers stay pure (detect structure), holder enforces policy.

That said — if there's a path to making this work through the reasoning parsers cleanly, I'm open to it. Is there an existing interface in the parsers we could hook into, or would this require a new parser API? Happy to explore that direction before merging if you think it's the right long-term home.

In the meantime, rebased on main after #43757 merged and resolved the class name conflict (test class renamed to `TestThinkingBudgetNaturalEndReentry` to coexist with #43757's `TestThinkingBudgetReentry`). Also clarifying: this PR *complements* #43757, not supersedes it — #43757 fixed forced-end re-entry, this fixes natural-end re-entry. Both are needed.

Validated on vLLM GPU instance:

| | Forced-end tests | Natural-end tests |
|---|---|---|
| Without patch | 2 fail | 4 fail |
| With patch | 2 pass ✓ | 6 pass ✓ |

**Comment by ashwing** [comment]
@simon-mo @rishitdholakia13 @llsj14  — this is the natural-end complement to #43757 which you approved/reviewed. #43757 fixed re-entry after *forced* `</think>` (budget exhausted); this fixes re-entry after *natural* `</think>` (model emits it before budget runs out). Same state machine, same stale-marker root cause, different exit path.

Rebased on main, conflicts resolved, 8/8 tests pass (both forced-end regressions and natural-end cases).

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
@bbrowning will you be designing this part in Q3?

**Comment by ashwing** [comment]
@simon-mo is this good to merge?

**Comment by bbrowning** [comment]
> @bbrowning will you be designing this part in Q3?

@chaunceyjiang has done some work in this area to use the parsers for thinking budget tracking, but it hasn't bubbled up to something I'm actively working on yet. I think a redesign is the only way to do this accurately across all models, as it's inherently the same logic we already have to do in the reasoning parsers to track when we're in reasoning and when we're not in reasoning for things like knowing when to apply structured output grammars or which content to parse as reasoning vs content vs tools.

If thinking token budgets are widely used, then I'd say yes we should tackle this in Q3.
